File: le-liaison/src/main/java/com/latticeengines/liaison/service/impl/QueryVDBImpl.java
package com.latticeengines.liaison.service.impl;

import com.latticeengines.liaison.exposed.exception.DefinitionException;
import com.latticeengines.liaison.exposed.service.Query;
import com.latticeengines.liaison.exposed.service.QueryColumn;
import com.latticeengines.liaison.service.impl.QueryColumnVDBImpl;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.regex.Matcher;


public class QueryVDBImpl extends Query {
	
	private static final Set<String> lasNotImpl = new HashSet<String>(
		Arrays.asList(
			"Fcn","AndNot","Meet","Join","Pushforward","Pullback","Relation","Copy"
			,"Port","Level","Conflict","PivotFunctions"
		)
	);
	
	private static final Set<String> laeNotImpl = new HashSet<String>(
		Arrays.asList(
			"Join","Meet","Copy","Relation","DivideIntoExistingNode","DivideNewNode","Port"
		)
	);
	
	private static final Pattern pattern_slq = Pattern.compile("^SpecLatticeQuery\\((LatticeAddressSet.*), SpecQueryNamedFunctions\\((.*)\\), (SpecQueryResultSet.*)\\)$");
	private static final Pattern pattern_las_atomic = Pattern.compile("^LatticeAddressSetPushforward\\((LatticeAddressExpression.*), LatticeAddressSetMeet\\((.*?)\\), (LatticeAddressExpressionAtomic\\(.*?\\))\\)$");
	private static final Pattern pattern_las_meet = Pattern.compile("^LatticeAddressSetPushforward\\((LatticeAddressExpression.*), LatticeAddressSetMeet\\((.*?)\\), LatticeAddressExpressionMeet\\((.*?)\\)\\)$");
	private static final Pattern pattern_filter_fcn = Pattern.compile("^(LatticeAddressSetFcn\\(LatticeFunction.*?LatticeAddressSet.*?\\))(, LatticeAddressSet.*|$)");
	private static final Pattern pattern_filter_las = Pattern.compile("^(LatticeAddressSet(.*?)\\((.*?)\\))(, LatticeAddressSet.*|$)");
	private static final Pattern pattern_entity_lae = Pattern.compile("^(LatticeAddressExpression(.*?)\\((.*?)\\))(, LatticeAddressExpression.*|$)");
	private static final Pattern pattern_sqnf = Pattern.compile("^(SpecQueryNamedFunction.*?)( SpecQueryNamedFunction.*|$)");

	public QueryVDBImpl( String name, String definition ) throws DefinitionException {
		super( name, definition );
	}
	
	public QueryVDBImpl( QueryVDBImpl other ) {
		super( other );
	}
	
	public QueryColumn createColumn( QueryColumn qc ) {
		QueryColumnVDBImpl vdbcol = (QueryColumnVDBImpl)qc;
		return new QueryColumnVDBImpl( vdbcol );
	}
	
	public void initFromDefinition( String name, String defn ) throws DefinitionException {
		
		List<QueryColumn> columns = new ArrayList<>();
		List<String> filterDefns = new ArrayList<>();
		List<String> entityDefns = new ArrayList<>();
		String resultSetDefn = "NOT SET";
		
		String las_spec;
		String sqnf;
		String lasm = "NOT SET";
		String lae = "NOT SET";
		boolean isMatchedTopLevel = Boolean.FALSE;
		
		Matcher m_slq = pattern_slq.matcher( defn );
		if( !m_slq.matches() ) {
			throw new DefinitionException( String.format("Maude definition (SpecLatticeQuery) cannot be interpretted: %s",defn) );
		}
		
		las_spec = m_slq.group(1);
		sqnf = m_slq.group(2);
		resultSetDefn = m_slq.group(3);
		
		if( !isMatchedTopLevel ) {
			Matcher m_las_atomic = pattern_las_atomic.matcher( las_spec );
			if( m_las_atomic.matches() ) {
				lasm = m_las_atomic.group(2);
				lae = "(" + m_las_atomic.group(3) + ")";
				isMatchedTopLevel = Boolean.TRUE;
			}
		}
		
		if( !isMatchedTopLevel ) {
			Matcher m_las_meet = pattern_las_meet.matcher( las_spec );
			if( m_las_meet.matches() ) {
				lasm = m_las_meet.group(2);
				lae = m_las_meet.group(3);
				isMatchedTopLevel = Boolean.TRUE;
			}
		}
		
		if( !isMatchedTopLevel ) {
			throw new DefinitionException( String.format("Maude definition (LatticeAddressSet) cannot be interpretted: %s",las_spec) );
		}
		
		if( !lasm.equals("empty") ) {
			lasm = lasm.substring(1,lasm.length()-1);
			while(Boolean.TRUE) {
				
				String filterspec = "";
				String remainingspec = "";
				boolean isMatched = Boolean.FALSE;
				
				if( !isMatched ) {
					Matcher m_filter_fcn = pattern_filter_fcn.matcher( lasm );
					if( m_filter_fcn.matches() ) {
						filterspec = m_filter_fcn.group(1);
						remainingspec = m_filter_fcn.group(2);
						isMatched = Boolean.TRUE;
					}
				}
				
				if( !isMatched ) {
					Matcher m_filter_las = pattern_filter_las.matcher( lasm );
					if( m_filter_las.matches() ) {
						if( lasNotImpl.contains(m_filter_las.group(2)) ) {
							throw new DefinitionException( String.format("Maude definition LatticeAddressSet%s not implemented",m_filter_las.group(2)) );
						}
						filterspec = m_filter_las.group(1);
						remainingspec = m_filter_las.group(4);
						isMatched = Boolean.TRUE;
					}
				}
				
				if( !isMatched ) {
					throw new DefinitionException( String.format("Maude definition (LatticeAddressSet) cannot be interpretted: %s",lasm) );
				}
				
				filterDefns.add( filterspec );
				if( remainingspec.equals("") ) {
					break;
				}
				else {
					lasm = remainingspec.substring(2);
				}
			}
		}
		
		if( !lae.equals("empty") ) {
			lae = lae.substring(1,lae.length()-1);
			while(Boolean.TRUE) {
				
				String entityspec = "";
				String remainingspec = "";
				boolean isMatched = Boolean.FALSE;
				
				if( !isMatched ) {
					Matcher m_entity_lae = pattern_entity_lae.matcher( lae );
					if( m_entity_lae.matches() ) {
						if( laeNotImpl.contains(m_entity_lae.group(2)) ) {
							throw new DefinitionException( String.format("Maude definition LatticeAddressExpression%s not implemented",m_entity_lae.group(2)) );
						}
						entityspec = m_entity_lae.group(1);
						remainingspec = m_entity_lae.group(4);
						isMatched = Boolean.TRUE;
					}
				}
				
				if( !isMatched ) {
					throw new DefinitionException( String.format("Maude definition (LatticeAddressExpression) cannot be interpretted: %s",lae) );
				}
				
				entityDefns.add( entityspec );
				if( remainingspec.equals("") ) {
					break;
				}
				else {
					lae = remainingspec.substring(2);
				}
			}
		}
		
		while(Boolean.TRUE) {
			Matcher m_sqnf = pattern_sqnf.matcher( sqnf );
			if( !m_sqnf.matches() ) {
				throw new DefinitionException( String.format("Maude definition (SpecQueryNamedFunction) cannot be interpretted: %s",sqnf) );
			}
			columns.add( new QueryColumnVDBImpl(m_sqnf.group(1)) );
			if( m_sqnf.group(2).equals("") ) {
				break;
			}
			else {
				sqnf = m_sqnf.group(2).substring(1);
			}
		}
		
		initFromValues( name, columns, filterDefns, entityDefns, resultSetDefn );
	}
	
	public String definition() {
		StringBuilder defn = new StringBuilder(100000);
		String sep = "";
		
		defn.append(               "SpecLatticeNamedElement("                    );
		defn.append(                 "SpecLatticeQuery("                         );
		defn.append(                   "LatticeAddressSetPushforward("           );
		
		if( getFilterDefns().size() == 0 ) {
			defn.append(                 "LatticeAddressExpressionFromLAS(LatticeAddressSetMeet(empty))");
			defn.append(               ", LatticeAddressSetMeet(empty)"          );
		}
		else {
			sep = "";
			StringBuilder tmplas = new StringBuilder(10000);
			tmplas.append(               "LatticeAddressSetMeet(("               );
			for( String f : getFilterDefns() ) {
				tmplas.append( sep );
				tmplas.append( f );
				sep = ", ";
			}
			tmplas.append(               "))"                                    );
			defn.append( String.format(  "LatticeAddressExpressionFromLAS(%s)",tmplas.toString()) );
			defn.append( String.format(", %s",tmplas.toString())                 );
		}
		
		if( getEntityDefns().size() == 0 ) {
			defn.append(               ", LatticeAddressExpressionMeet(empty)"   );
		}
		else {
			sep = "";
			defn.append(               ", LatticeAddressExpressionMeet(("        );
			for( String e : getEntityDefns() ) {
				defn.append( sep );
				defn.append( e );
				sep = ", ";
			}
			defn.append(                 "))"                                    );
		}
		
		defn.append(                   ")"                                       );
		defn.append(                 ", SpecQueryNamedFunctions("                );
		
		sep = "";
		for( QueryColumn c : getColumns() ) {
			defn.append( sep );
			defn.append( c.definition() );
			sep = " ";
		}
		
		defn.append(                   ")"                                       );
		defn.append(                 ", " + getResultSetDefn()                   );
		defn.append(                 ")"                                         );
		defn.append( String.format(", ContainerElementName(\"%s\")",getName())   );
		defn.append(               ")"                                           );
		
		return defn.toString();
	}
}

File: le-liaison/src/main/python/liaison/QueryVDBImpl.py

#
# $LastChangedBy$
# $LastChangedDate$
# $Rev$
#

import re
from .exceptions import MaudeStringError, ExpressionNotImplemented
from .Query import Query
from .QueryColumnVDBImpl import QueryColumnVDBImpl
from .QueryEntityVDBImpl import QueryEntityVDBImpl
from .QueryFilterVDBImpl import QueryFilterVDBImpl


class QueryVDBImpl(Query):

    ## This is the default constructor

    def __init__(self, name, columns, filters = [], entities = [], resultSet = None):

        self.initFromValues(name, columns, filters, entities, resultSet)
    

## This is the pythonic way to create multiple constructors.  The call is
## q = QueryVDBImpl.InitFromDefn( defn )

    @classmethod
    def initFromDefn(cls, name, defn):

        s1 = re.search( '^SpecLatticeQuery\((LatticeAddressSet.*), SpecQueryNamedFunctions\((.*)\), (SpecQueryResultSet.*)\)$', defn )
        if not s1:
            raise MaudeStringError( defn )

        las_spec  = s1.group(1)

        lasm = ''
        lae = ''
        isMatchedTopLevel = False

        if not isMatchedTopLevel:
            s3 = re.search( '^LatticeAddressSetPushforward\(LatticeAddressExpressionFromLAS\(LatticeAddressSetMeet\((.*?)\)\), LatticeAddressSetMeet\((.*?)\), LatticeAddressExpressionMeet\((.*)\)\)$', las_spec )
            if s3:
                lasm = s3.group(2)
                lae = s3.group(3)
                isMatchedTopLevel = True

        if not isMatchedTopLevel:
            s3 = re.search( '^LatticeAddressSetPushforward\(LatticeAddressExpressionFromLAS\(LatticeAddressSetMeet\((.*?)\)\), LatticeAddressSetMeet\((.*?)\), (LatticeAddressExpressionAtomic\(.*\))\)$', las_spec )
            if s3:
                lasm = s3.group(2)
                lae = '({0})'.format( s3.group(3) )
                isMatchedTopLevel = True

        if not isMatchedTopLevel:
            s3 = re.search( '^LatticeAddressSetPushforward\(LatticeAddressExpressionFromLAS\((LatticeAddressSetIdentifier\(.*?\))\), (LatticeAddressSetIdentifier\(.*?\)), LatticeAddressExpressionMeet\((.*)\)\)$', las_spec )
            if s3:
                lasm = '({0})'.format( s3.group(2) )
                lae = s3.group(3)
                isMatchedTopLevel = True

        if not isMatchedTopLevel:
            s3 = re.search( '^LatticeAddressSetPushforward\(LatticeAddressExpressionFromLAS\((LatticeAddressSetIdentifier\(.*?\))\), (LatticeAddressSetIdentifier\(.*?\)), (LatticeAddressExpressionAtomic\(.*\))\)$', las_spec )
            if s3:
                lasm = '({0})'.format( s3.group(2) )
                lae = '({0})'.format( s3.group(3) )
                isMatchedTopLevel = True

        if not isMatchedTopLevel:
            s3 = re.search( '^LatticeAddressSetPushforward\(LatticeAddressExpressionMeet\((.*?)\), LatticeAddressSetMeet\((.*?)\), LatticeAddressExpressionMeet\((.*)\)\)$', las_spec )
            if s3:
                lasm = s3.group(2)
                lae = s3.group(3)
                isMatchedTopLevel = True

        if not isMatchedTopLevel:
            s3 = re.search( '^LatticeAddressSetPushforward\(LatticeAddressExpressionMeet\((.*?)\), LatticeAddressSetMeet\((.*?)\), (LatticeAddressExpressionAtomic\(.*\))\)$', las_spec )
            if s3:
                lasm = s3.group(2)
                lae = '({0})'.format( s3.group(3) )
                isMatchedTopLevel = True

        if not isMatchedTopLevel:
            raise MaudeStringError( las_spec )

        
        filters = []
        entities = []
        
        if lasm != 'empty':

            lasm = lasm[1:-1]

            while True:
                
                filterspec = ''
                remainingspec = ''
                isMatched = False

                if not isMatched:
                    s31 = re.search( '^(LatticeAddressSetFcn\(LatticeFunction.*?LatticeAddressSet.*?\))(, LatticeAddressSet.*|$)', lasm )
                    if s31:
                        filterspec = s31.group(1)
                        remainingspec = s31.group(2)
                        isMatched = True

                if not isMatched:
                    s31 = re.search( '^(LatticeAddressSet(.*?)\((.*?)\))(, LatticeAddressSet.*|$)', lasm )
                    if not s31:
                        raise MaudeStringError( lasm )

                    if s31.group(2) in set(['Fcn','AndNot','Meet','Join','Pushforward','Pullback','Relation','Copy','Port','Level','Conflict','PivotFunctions']):
                        raise ExpressionNotImplemented( 'LatticeAddressSet{0}: {1}'.format(s31.group(2),lasm) )

                    filterspec = s31.group(1)
                    remainingspec = s31.group(4)
                    isMatched = True

                filters.append( QueryFilterVDBImpl(filterspec) )
                if remainingspec == '':
                    break
                else:
                    # Strip off the ', ' at the beginning
                    lasm = remainingspec[2:]

        if lae != 'empty':

            lae = lae[1:-1]

            while True:

                entityspec = ''
                remainingspec = ''
                isMatched = False

                if not isMatched:
                    s32 = re.search( '^(LatticeAddressExpression(.*?)\((.*?)\))(, LatticeAddressExpression.*|$)', lae )
                    if not s32:
                        raise MaudeStringError( lae )

                    if s32.group(2) in set(['Join','Meet','Copy','Relation','DivideIntoExistingNode','DivideNewNode','Port']):
                        raise ExpressionNotImplemented( 'LatticeAddressExpression{0}'.format(s32.group(2)) )

                    entityspec = s32.group(1)
                    remainingspec = s32.group(4)
                    isMatched = True

                entities.append( QueryEntityVDBImpl(entityspec) )
                if remainingspec == '':
                    break
                else:
                    # Strip off the ', ' at the beginning
                    lae = remainingspec[2:]

        resultSet = s1.group(3)

        columns = []
        sqnf = s1.group(2)

        ## Iterate on the comma-separated SpecLatticeSourceTableColumn specs
        while True:
            s2 = re.search( '^(SpecQueryNamedFunction.*?)( SpecQueryNamedFunction.*|$)', sqnf )
                
            if not s2:
                raise MaudeStringError( sqnf )

            c = QueryColumnVDBImpl.initFromDefn( s2.group(1) )
            columns.append( c )

            if s2.group(2) == '':
                break
            else:
                # Strip off the ' ' at the beginning
                sqnf = s2.group(2)[1:]

        return cls(name, columns, filters, entities, resultSet)


    def initFromValues(self, name, columns, filters, entities, resultSet = None):

        super(QueryVDBImpl, self).initFromValues(name, columns, filters, entities, resultSet)

        if resultSet is not None:
            self.resultSet_ = resultSet

        else:
            self.resultSet_ = 'SpecQueryResultSetAll'


    def SpecLatticeNamedElements(self):

        defn =  'SpecLatticeNamedElements(('
        defn +=   'SpecLatticeNamedElement('
        defn +=     'SpecLatticeQuery('
        defn +=       'LatticeAddressSetPushforward('

        if not self.filters_:
            defn +=       'LatticeAddressExpressionFromLAS(LatticeAddressSetMeet(empty))'
            defn +=     ', LatticeAddressSetMeet(empty)'
        else:
            sep = ''
            tmplas =      'LatticeAddressSetMeet(('
            for f in self.filters_:
                tmplas += sep
                tmplas += f.definition()
                sep = ', '
            tmplas +=     '))'
            defn +=       'LatticeAddressExpressionFromLAS({0})'.format( tmplas )
            defn +=     ', {0}'.format( tmplas )

        if not self.entities_:
            defn +=     ', LatticeAddressExpressionMeet(empty)'
        else:
            sep = ''
            defn +=     ', LatticeAddressExpressionMeet(('
            for e in self.entities_:
                defn += sep
                defn += e.definition()
                sep = ', '
            defn +=       '))'

        defn +=       ')'
        defn +=     ', SpecQueryNamedFunctions('

        sep = ''
        for c in self.columns_:
            defn += sep
            defn += c.definition()
            sep = ' '

        defn +=       ')'
        defn +=     ', ' + self.resultSet_
        defn +=     ')'
        defn +=   ', ContainerElementName(\"{0}\")'.format( self.getName() )
        defn +=   ')'
        defn += '))'

        return defn
